5/testing-genesis.md

```diff
@@ -0,0 +1,147 @@
+Testing Genesis
+===
+
+Right about meow, you may have noticed that the test setup can get quite tedious,
+i.e. having to create new kitties for each unit test.
+
+Wouldn't it be great to deploy our blockchain with some kitties already minted in the genesis block?
+
+Substrate lets you deploy your chain with preconfigured storage through a genesis configuration.
+
+In this section, we'll walk you through: 
+- Extending `decl_storage` to add extra genesis data
+- Mocking the genesis configuration in during testing
+- Testing that we have the correct genesis set up
+
+## Add Extra Genesis
+
+Let's start by importing some functions and types we'll be using to configure genesis.
+
+**substratekitties<span>.</span>rs**
+```rust
+use runtime_io::{with_storage, StorageOverlay, ChildrenStorageOverlay};
+```
+Notably:
+- `with_storage`: is a function that takes a closure with global functions. This function lets you populate the runtime storage.
+- `StorageOverlay`: is a set of key value pairs for storage. This represents where your storage resides.
+
+Meow, let's configure our chain's genesis to accept some initial kitties, with predefined DNA & value.
+And let's assign those kitties to designed owners.
+
+Inside the `decl_storage` scope, create a struct called `add_extra_genesis` with the following implementation. 
+
+```rust
+decl_storage! {
+    trait Store for Module<T: Trait> as KittyStorage { ... }
+
+    add_extra_genesis {
+        config(kitties): Vec<(T::AccountId, T::Hash, T::Balance)>;
+        
+        build(|storage: &mut StorageOverlay, _: &mut ChildrenStorageOverlay, config: &GenesisConfig<T>| {
+            with_storage(storage, || {
+                for &(ref acct, hash, balance) in &config.kitties {
+
+                    let k = Kitty {
+                                id: hash,
+                                dna: hash,
+                                price: balance,
+                                gen: 0
+                            };
+                
+                    let _ = <Module<T>>::mint(acct.clone(), hash, k);
+                }
+            });
+        });
+    }
+}
+```
+The `config` attribute will expect a `kitties` config value, which will be a vector of tuples with types: 
+- AccountId: the kitty owner
+- Hash: the kitty id/dna
+- Balance: the kitty's initial value
+
+`build` subsequently uses the `kitties` configuration to build the storage itself.
+In this implementation, it iterates over `kitties` and use the `mint` function to conveniently update
+the rest of the storage items.
+
+> **Note**: For this part to work, delete (or comment out) the deposit_event line in `mint()`, as `add_extra_genesis`
```

Comment:
This is a strange requirement, and ultimately not very helpful for someone who wants to write tests using this pattern... what are the solutions?

Reply:
I was expecting something like this: https://github.com/paritytech/substrate/blob/master/srml/staking/src/mock.rs#L124

Reply:
TBH I do not know much about `add_extra_genesis` and this section does not explain it well

Reply:
So we have 3 paths forward for this section: 

1. Marking storage values with `config()`, then configuring each storage value 1 by 1, as @shawntabrizi indicated. This is easier to understand and we have a few tutorials that already cover this. e.g. [1](https://docs.substrate.dev/docs/building-the-substrate-tcr-runtime#section-storage-declaration-using-the-decl_storage-macro) and [2](https://substrate.recipes/testing/index.html)

2. Building with build() inside each storage item declaration. I implemented this as well, before discovering the 3rd approach. i.e.
```rust
        Kitties get(kitty) build(|config: &GenesisConfig<T>| {
            config.kitties.iter()
                .map(|i| ( i.1, Kitty { id: i.1, dna: i.1, price: i.2, gen: 0, } ))
                .collect::<Vec<_>>()
        }): map T::Hash => Kitty<T::Hash, T::Balance>;

        KittyOwner get(owner_of) build(|config: &GenesisConfig<T>| {
            config.kitties.iter().cloned() //address this
                .map(|i| (i.1, i.0) )
                .collect::<Vec<_>>()
        }): map T::Hash => Option<T::AccountId>;
```

3. Waiting to backport latest master features to V1.0, so we can build storage inside `add_extra_genesis` (given a much better explanation 😄 ). 

I still advocate for 3 since Kitties' storage values are interdependent and this feels more elegant, less manual, less lines. 

But I do see value in also teaching (1) and (2), which are common patterns as well. Thoughts @shawntabrizi and @kianenigma ?